### PR TITLE
Support dynamic mhlo.torch_index_select and add test.

### PR DESCRIPTION
**Pull request by shanshanpt**
Support dynamic mhlo.torch_index_select and add test.

### CODE REVIEW

**Comment by shanshanpt** [comment]
Add ShapeBuilder of mhlo TorchIndexSelectOp,  that support dynamic shape TorchIndexSelectOp.

File: iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp [hunk 1]
```diff
@@ -308,6 +308,128 @@ Value rewriteSelectOp(RankedShapeType resultShape, SelectOp selectOp,
...
+                              OpBuilder &builder) {
+  if (!torchIndexSelectOp) return nullptr;
+  auto loc = torchIndexSelectOp.getLoc();
+
+  int64_t resultShapeRank = resultShape.getRank();
+  auto paramsType =
+      torchIndexSelectOp.input().getType().cast<RankedTensorType>();
+  auto indicesType =
+      torchIndexSelectOp.index().getType().cast<RankedTensorType>();
+  if (!paramsType || !indicesType) {
```
Comment (stellaraccident) [inline iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp:322]: Since you are checking for nullptr, what you want above is dyncast<>, not cast<>. cast will never return nullptr and will assert fail on a mismatched type.

File: iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp [hunk 2]
```diff
@@ -308,6 +308,128 @@ Value rewriteSelectOp(RankedShapeType resultShape, SelectOp selectOp,
...
+    return nullptr;
+  }
+
+  auto axis = torchIndexSelectOp.dim();
+  auto batchDim = torchIndexSelectOp.batch_dims();
+  int64_t paramsRank = paramsType.getRank();
+  int64_t indicesRank = indicesType.getRank();
+
+  std::vector<int64_t> shape(paramsType.getShape());
+  int64_t axisValue = *(axis.getRawData());
```
Comment (stellaraccident) [inline iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp:332]: I think you want axis.getSExtValue() here and on the next line.

File: iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp [hunk 3]
```diff
@@ -308,6 +308,128 @@ Value rewriteSelectOp(RankedShapeType resultShape, SelectOp selectOp,
...
+      GENERATE_RANKED_DIM_OP(indicesShapeValue, i);
+    }
+    // params.shape[indicesRank-batchDim+axisValue:]
+    // resultShapeIndex == indicesRank-batchDim+axisValue
+    for (int64_t i = resultShapeIndex; i < resultShapeRank; ++i) {
+      if (!resultShape.isDimDynamic(resultShapeIndex++)) continue;
+      GENERATE_RANKED_DIM_OP(paramsShapeValue, i);
+    }
+  } else {
+    if (resultShapeRank != paramsRank + indicesRank - 1 - batchDimValue) {
```
Comment (stellaraccident) [inline iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp:405]: Nit: can you add parens around the rhs?

File: iree/test/e2e/regression/dynamic_torch_index_select_high_rank.mlir [hunk 4]
```diff
@@ -0,0 +1,58 @@
+// RUN: iree-run-mlir %s -iree-hal-target-backends=llvm-ir -input-value="2x2xi32=[6, 7] [8, 9]" -input-value="2x2x2x2xi32=[[[0, 1] [1, 0]] [[0, 0] [1, 1]]] [[[1, 1] [0, 0]] [[0, 1] [1, 0]]]" | IreeFileCheck %s
```
Comment (stellaraccident) [inline iree/test/e2e/regression/dynamic_torch_index_select_high_rank.mlir:1]: Can you point me to (or add a test) where we are testing the wrap-around (negative index) behavior?

Comment (shanshanpt) [reply iree/test/e2e/regression/dynamic_torch_index_select_high_rank.mlir:1]: In tensorflow, axis and batch_dims can be negative. 
But the function `TorchIndexSelectOpConversion` in HLOToLinalgOnBuffers.cpp disable the nagtive, so modify the code and add the negative case.

File: iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp (hunk 3 above)
Comment (shanshanpt) [reply iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp:405]: done

File: iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp (hunk 2 above)
Comment (shanshanpt) [reply iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp:332]: done

File: iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp (hunk 1 above)
Comment (shanshanpt) [reply iree/compiler/Dialect/Shape/Plugins/XLA/XlaHloShapeBuilder.cpp:322]: done

**Comment by shanshanpt** [comment]
@googlebot I signed it!
